### DISCUSSION TITLE
Question about PR #802 - msg_incompat_call_details

### DISCUSSION THREAD

**Discussion by tonygermano** [mozilla/rhino#1592] [General]
I understand the purpose of #802, and I think it is valuable for troubleshooting Rhino issues, but I wonder if it is inappropriately leaking implementation details to the javascript environment in some cases and a Context flag should control whether or not to display the java class names. 

For example, in Rhino I get:
```javascript
js> Number.prototype.toString.call(Boolean)
js: uncaught JavaScript runtime exception: TypeError: Method "toString" called on incompatible
 object (org.mozilla.javascript.IdFunctionObjectES6 is not an instance of
 org.mozilla.javascript.NativeNumber).
```

In Chrome I get:
```javascript
>  Number.prototype.toString.call(Boolean)
<- Uncaught TypeError: Number.prototype.toString requires that 'this' be a Number
```

I think a message like what Chrome provides would be much more useful than the message was prior to this PR without leaking java class names.

**Comment by tonygermano** [comment]
I came across this while looking into why https://github.com/tc39/test262/blob/main/test/built-ins/BigInt/prototype/toString/prototype-call.js was not passing.

In Chrome:
```javascript
>  Number.prototype.toString.call(Number.prototype)
<- '0'
>  BigInt.prototype.toString.call(BigInt.prototype)
<- Uncaught TypeError: BigInt.prototype.toString requires that 'this' be a BigInt
```

In Rhino:
```javascript
js> Number.prototype.toString.call(Number.prototype)
0
js> BigInt.prototype.toString.call(BigInt.prototype)
0
```

According to the spec,
- `Number.prototype` is itself a Number object; it has a [[NumberData]] internal slot with the value +0𝔽.
- `BigInt.prototype`is not a BigInt object; it does not have a [[BigIntData]] internal slot.

But the problem is that `BigInt.prototype` is an instance of `NativeBigInt` because I think that's how `IdScriptableObject` works, and I assume `NativeNumber` was used as a template for `NativeBigInt`. So, the type check actually needs to check that it is an instanceof `NativeBigInt` AND it is NOT the `BigInt.prototype` instance. In the latter case throwing the java name would not make sense, because it is an instance of `NativeBigInt`.

**Comment by p-bakker** [comment]
I think you're right that the impl. details should not leak into the exception like that. Case and possibly a PR?

As for the wrong behaviour with BigInt.prototype.toString: don't think it's an inherent probleem in IdScriptable, but rather an issue in the BigInt impl. Case?

While on the topic of BigInt: a lot of tests from test262 fail when the feature under testing is used icw BigInt. Wood be great to get some insight into if there's a common denominator between all or many of those failures

**Reply by tonygermano** [reply]
The only reason that I said that about IdScriptableObject is because I think it's the case for all of the other classes representing primitive types that the prototype object is an instance of the class, but not for BigInt. All of the methods for BigInt.prototype are implemented inside of NativeBigInt::execIdCall. The constructor methods are also implemented in the same function.

**Reply by tonygermano** [reply]
Other examples of prototype java classes. The leaky error message is useful here:
```javascript
js> BigInt.prototype.toString.call(Array.prototype)
js: uncaught JavaScript runtime exception: TypeError: Method "toString" called on incompatible object (org.mozilla.javascript.NativeArray is not an instance of org.mozilla.javascript.NativeBigInt).
        at (unknown):3

js> BigInt.prototype.toString.call(Symbol.prototype)
js: uncaught JavaScript runtime exception: TypeError: Method "toString" called on incompatible object (org.mozilla.javascript.NativeSymbol is not an instance of org.mozilla.javascript.NativeBigInt).
        at (unknown):4

js> BigInt.prototype.toString.call(Uint8Array.prototype)
js: uncaught JavaScript runtime exception: TypeError: Method "toString" called on incompatible object (org.mozilla.javascript.typedarrays.NativeUint8Array is not an instance of org.mozilla.javascript.NativeBigInt).
        at (unknown):6

js> BigInt.prototype.toString.call(Date.prototype)
js: uncaught JavaScript runtime exception: TypeError: Method "toString" called on incompatible object (org.mozilla.javascript.NativeDate is not an instance of org.mozilla.javascript.NativeBigInt).
        at (unknown):7
```
